File: internal/server/middleware.go
package server

import (
	"context"
	"log/slog"
	"net/http"
	"time"

	"github.com/google/uuid"

	"github.com/tjfontaine/polyglot-llm-gateway/internal/auth"
)

// RequestIDKey is the context key for request IDs
type contextKey string

const RequestIDKey contextKey = "request_id"

// logFieldsKey identifies request-scoped logging fields.
type logFieldsKey struct{}

// RequestIDMiddleware adds a unique request ID to each request
func RequestIDMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		requestID := uuid.New().String()
		ctx := context.WithValue(r.Context(), RequestIDKey, requestID)
		w.Header().Set("X-Request-ID", requestID)
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}

// LoggingMiddleware logs HTTP requests with structured logging
func LoggingMiddleware(logger *slog.Logger) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			start := time.Now()

			// Attach mutable log fields map to context for handlers to enrich
			fields := make(map[string]string)
			ctxWithFields := context.WithValue(r.Context(), logFieldsKey{}, fields)

			// Wrap response writer to capture status code
			wrapped := &responseWriter{ResponseWriter: w, statusCode: http.StatusOK}

			// Get request ID from context
			requestID, _ := r.Context().Value(RequestIDKey).(string)

			// Log request start
			logger.Info("request started",
				slog.String("request_id", requestID),
				slog.String("method", r.Method),
				slog.String("path", r.URL.Path),
				slog.String("remote_addr", r.RemoteAddr),
			)

			next.ServeHTTP(wrapped, r.WithContext(ctxWithFields))

			// Log request completion
			duration := time.Since(start)
			attrs := []slog.Attr{
				slog.String("request_id", requestID),
				slog.String("method", r.Method),
				slog.String("path", r.URL.Path),
				slog.Int("status", wrapped.statusCode),
				slog.Duration("duration", duration),
			}

			if len(fields) > 0 {
				for k, v := range fields {
					attrs = append(attrs, slog.String(k, v))
				}
			}

			logger.LogAttrs(ctxWithFields, slog.LevelInfo, "request completed", attrs...)
		})
	}
}

// responseWriter wraps http.ResponseWriter to capture status code
type responseWriter struct {
	http.ResponseWriter
	statusCode int
}

func (rw *responseWriter) WriteHeader(code int) {
	rw.statusCode = code
	rw.ResponseWriter.WriteHeader(code)
}

// Flush forwards Flush to the underlying ResponseWriter if it supports http.Flusher,
// preserving streaming support (e.g., for SSE).
func (rw *responseWriter) Flush() {
	if f, ok := rw.ResponseWriter.(http.Flusher); ok {
		f.Flush()
	}
}

// TimeoutMiddleware enforces request timeouts
func TimeoutMiddleware(timeout time.Duration) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			ctx, cancel := context.WithTimeout(r.Context(), timeout)
			defer cancel()
			next.ServeHTTP(w, r.WithContext(ctx))
		})
	}
}

// AddLogField attaches a key/value to the request-scoped log fields map so LoggingMiddleware can emit it.
// It is safe to call multiple times. No-op if middleware isn't present.
func AddLogField(ctx context.Context, key, value string) {
	if value == "" {
		return
	}
	if fields, ok := ctx.Value(logFieldsKey{}).(map[string]string); ok {
		fields[key] = value
	}
}

// AddError attaches an error message to the request-scoped log fields map so it
// appears in the structured request log emitted by LoggingMiddleware. No-op if
// middleware isn't present or err is nil.
func AddError(ctx context.Context, err error) {
	if err == nil {
		return
	}
	AddLogField(ctx, "error", err.Error())
}

// AuthMiddleware validates API keys and injects tenant context
func AuthMiddleware(authenticator *auth.Authenticator) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Extract API key from Authorization header
			apiKey := r.Header.Get("Authorization")
			if apiKey == "" {
				http.Error(w, "Missing Authorization header", http.StatusUnauthorized)
				return
			}

			// Remove "Bearer " prefix if present
			if len(apiKey) > 7 && apiKey[:7] == "Bearer " {
				apiKey = apiKey[7:]
			}

			// Validate API key and get tenant
			tenant, err := authenticator.ValidateAPIKey(apiKey)
			if err != nil {
				http.Error(w, "Invalid API key", http.StatusUnauthorized)
				return
			}

			// Inject tenant into context
			ctx := context.WithValue(r.Context(), "tenant", tenant)
			next.ServeHTTP(w, r.WithContext(ctx))
		})
	}
}

// rateLimitContextKey is the context key for rate limit info
type rateLimitContextKey struct{}

// SetRateLimits stores rate limit info in context for the middleware to write as headers.
func SetRateLimits(ctx context.Context, rl *RateLimitInfo) context.Context {
	return context.WithValue(ctx, rateLimitContextKey{}, rl)
}

// RateLimitInfo contains normalized rate limit information.
// This is a copy of domain.RateLimitInfo to avoid import cycles.
type RateLimitInfo struct {
	RequestsLimit     int
	RequestsRemaining int
	RequestsReset     string
	TokensLimit       int
	TokensRemaining   int
	TokensReset       string
}

// RateLimitNormalizingMiddleware writes normalized rate limit headers to responses.
// It reads rate limit info from context (set by frontdoors after provider calls)
// and writes standardized x-ratelimit-* headers.
func RateLimitNormalizingMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Create a wrapper that will write headers after the handler sets rate limits
		wrapped := &rateLimitResponseWriter{
			ResponseWriter: w,
			request:        r,
		}
		next.ServeHTTP(wrapped, r)
	})
}

// rateLimitResponseWriter wraps ResponseWriter to write rate limit headers.
type rateLimitResponseWriter struct {
	http.ResponseWriter
	request       *http.Request
	wroteHeaders  bool
}

func (rw *rateLimitResponseWriter) WriteHeader(code int) {
	if !rw.wroteHeaders {
		rw.writeRateLimitHeaders()
		rw.wroteHeaders = true
	}
	rw.ResponseWriter.WriteHeader(code)
}

func (rw *rateLimitResponseWriter) Write(b []byte) (int, error) {
	if !rw.wroteHeaders {
		rw.writeRateLimitHeaders()
		rw.wroteHeaders = true
	}
	return rw.ResponseWriter.Write(b)
}

func (rw *rateLimitResponseWriter) writeRateLimitHeaders() {
	rl, ok := rw.request.Context().Value(rateLimitContextKey{}).(*RateLimitInfo)
	if !ok || rl == nil {
		return
	}

	h := rw.Header()

	// Write normalized rate limit headers
	// Standard format: x-ratelimit-{limit|remaining|reset}-{requests|tokens}
	if rl.RequestsLimit > 0 {
		h.Set("x-ratelimit-limit-requests", itoa(rl.RequestsLimit))
	}
	if rl.RequestsRemaining > 0 {
		h.Set("x-ratelimit-remaining-requests", itoa(rl.RequestsRemaining))
	}
	if rl.RequestsReset != "" {
		h.Set("x-ratelimit-reset-requests", rl.RequestsReset)
	}

	if rl.TokensLimit > 0 {
		h.Set("x-ratelimit-limit-tokens", itoa(rl.TokensLimit))
	}
	if rl.TokensRemaining > 0 {
		h.Set("x-ratelimit-remaining-tokens", itoa(rl.TokensRemaining))
	}
	if rl.TokensReset != "" {
		h.Set("x-ratelimit-reset-tokens", rl.TokensReset)
	}
}

// itoa converts int to string without importing strconv
func itoa(i int) string {
	if i == 0 {
		return "0"
	}
	
	negative := i < 0
	if negative {
		i = -i
	}
	
	var buf [20]byte
	pos := len(buf)
	
	for i > 0 {
		pos--
		buf[pos] = byte('0' + i%10)
		i /= 10
	}
	
	if negative {
		pos--
		buf[pos] = '-'
	}
	
	return string(buf[pos:])
}

// Flush forwards Flush to the underlying ResponseWriter if it supports http.Flusher.
func (rw *rateLimitResponseWriter) Flush() {
	if f, ok := rw.ResponseWriter.(http.Flusher); ok {
		f.Flush()
	}
}

File: ts/packages/gateway-core/src/http/middleware.ts
/**
 * HTTP middleware for gateway requests.
 *
 * Provides request ID, logging, timeout, and rate limit normalization
 * middleware that work with Web standard Request/Response.
 *
 * @module http/middleware
 */

import type { Logger } from '../utils/logging.js';
import { randomUUID } from '../utils/crypto.js';

// ============================================================================
// Context Types
// ============================================================================

/**
 * Request context with middleware-injected values.
 */
export interface RequestContext {
    /** Unique request ID. */
    requestId: string;

    /** Request start time. */
    startTime: number;

    /** Custom log fields added during request handling. */
    logFields: Map<string, string>;

    /** Rate limit info to write as response headers. */
    rateLimits?: HttpRateLimitInfo | undefined;
}

/**
 * Rate limit information for response headers.
 * Named HttpRateLimitInfo to avoid conflict with domain RateLimitInfo.
 */
export interface HttpRateLimitInfo {
    requestsLimit?: number | undefined;
    requestsRemaining?: number | undefined;
    requestsReset?: string | undefined;
    tokensLimit?: number | undefined;
    tokensRemaining?: number | undefined;
    tokensReset?: string | undefined;
}

// ============================================================================
// Context Storage
// ============================================================================

/**
 * WeakMap to store request context without modifying Request object.
 */
const requestContexts = new WeakMap<Request, RequestContext>();

/**
 * Gets the request context for a request.
 */
export function getRequestContext(request: Request): RequestContext | undefined {
    return requestContexts.get(request);
}

/**
 * Gets or creates a request context.
 */
export function ensureRequestContext(request: Request): RequestContext {
    let ctx = requestContexts.get(request);
    if (!ctx) {
        ctx = {
            requestId: randomUUID(),
            startTime: Date.now(),
            logFields: new Map(),
        };
        requestContexts.set(request, ctx);
    }
    return ctx;
}

/**
 * Gets the request ID from a request.
 */
export function getRequestId(request: Request): string {
    return getRequestContext(request)?.requestId ?? '';
}

/**
 * Adds a log field to the request context.
 */
export function addLogField(request: Request, key: string, value: string): void {
    const ctx = getRequestContext(request);
    if (ctx && value) {
        ctx.logFields.set(key, value);
    }
}

/**
 * Adds an error to the request context log fields.
 */
export function addError(request: Request, error: Error): void {
    addLogField(request, 'error', error.message);
}

/**
 * Sets rate limit info in the request context.
 */
export function setRateLimits(request: Request, rateLimits: HttpRateLimitInfo): void {
    const ctx = getRequestContext(request);
    if (ctx) {
        ctx.rateLimits = rateLimits;
    }
}

// ============================================================================
// Middleware Types
// ============================================================================

/**
 * HTTP handler function type.
 */
export type HttpHandler = (request: Request) => Promise<Response>;

/**
 * HTTP middleware function type.
 */
export type HttpMiddleware = (handler: HttpHandler) => HttpHandler;

// ============================================================================
// Request ID Middleware
// ============================================================================

/**
 * Adds a unique request ID to each request.
 * Sets X-Request-ID response header.
 */
export function requestIdMiddleware(): HttpMiddleware {
    return (handler) => async (request) => {
        const ctx = ensureRequestContext(request);

        // Check for existing request ID in header
        const existingId = request.headers.get('x-request-id');
        if (existingId) {
            ctx.requestId = existingId;
        }

        const response = await handler(request);

        // Add request ID to response headers
        const newHeaders = new Headers(response.headers);
        newHeaders.set('X-Request-ID', ctx.requestId);

        return new Response(response.body, {
            status: response.status,
            statusText: response.statusText,
            headers: newHeaders,
        });
    };
}

// ============================================================================
// Logging Middleware
// ============================================================================

/**
 * Options for logging middleware.
 */
export interface LoggingMiddlewareOptions {
    /** Logger instance. */
    logger: Logger;

    /** Whether to log request bodies (default: false). */
    logRequestBody?: boolean;

    /** Paths to skip logging for (e.g., health checks). */
    skipPaths?: string[];
}

/**
 * Logs HTTP requests with structured logging.
 */
export function loggingMiddleware(options: LoggingMiddlewareOptions): HttpMiddleware {
    const { logger, skipPaths = ['/health', '/healthz', '/ready'] } = options;

    return (handler) => async (request) => {
        const url = new URL(request.url);

        // Skip logging for certain paths
        if (skipPaths.includes(url.pathname)) {
            return handler(request);
        }

        const ctx = ensureRequestContext(request);

        // Log request start
        logger.info('request started', {
            requestId: ctx.requestId,
            method: request.method,
            path: url.pathname,
        });

        let response: Response | undefined;
        let error: Error | undefined;

        try {
            response = await handler(request);
            return response;
        } catch (e) {
            error = e instanceof Error ? e : new Error(String(e));
            throw e;
        } finally {
            const duration = Date.now() - ctx.startTime;

            // Build log attributes
            const attrs: Record<string, unknown> = {
                requestId: ctx.requestId,
                method: request.method,
                path: url.pathname,
                status: response?.status ?? 500,
                durationMs: duration,
            };

            // Add custom log fields
            for (const [key, value] of ctx.logFields) {
                attrs[key] = value;
            }

            if (error) {
                attrs.error = error.message;
            }

            logger.info('request completed', attrs);
        }
    };
}

// ============================================================================
// Timeout Middleware
// ============================================================================

/**
 * Enforces request timeouts.
 * Uses AbortController to cancel long-running requests.
 */
export function timeoutMiddleware(timeoutMs: number): HttpMiddleware {
    return (handler) => async (request) => {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), timeoutMs);

        try {
            // Create a new request with the abort signal
            const newRequest = new Request(request.url, {
                method: request.method,
                headers: request.headers,
                body: request.body,
                signal: controller.signal,
            });

            // Copy context to new request
            const ctx = getRequestContext(request);
            if (ctx) {
                requestContexts.set(newRequest, ctx);
            }

            return await handler(newRequest);
        } catch (error) {
            if (error instanceof Error && error.name === 'AbortError') {
                return new Response(
                    JSON.stringify({
                        error: {
                            type: 'timeout',
                            message: `Request timed out after ${timeoutMs}ms`,
                        },
                    }),
                    {
                        status: 504,
                        headers: { 'Content-Type': 'application/json' },
                    },
                );
            }
            throw error;
        } finally {
            clearTimeout(timeoutId);
        }
    };
}

// ============================================================================
// Rate Limit Header Middleware
// ============================================================================

/**
 * Writes normalized rate limit headers to responses.
 */
export function rateLimitMiddleware(): HttpMiddleware {
    return (handler) => async (request) => {
        const response = await handler(request);
        const ctx = getRequestContext(request);

        if (!ctx?.rateLimits) {
            return response;
        }

        const rl = ctx.rateLimits;
        const newHeaders = new Headers(response.headers);

        // Write normalized rate limit headers
        if (rl.requestsLimit !== undefined) {
            newHeaders.set('x-ratelimit-limit-requests', String(rl.requestsLimit));
        }
        if (rl.requestsRemaining !== undefined) {
            newHeaders.set('x-ratelimit-remaining-requests', String(rl.requestsRemaining));
        }
        if (rl.requestsReset) {
            newHeaders.set('x-ratelimit-reset-requests', rl.requestsReset);
        }
        if (rl.tokensLimit !== undefined) {
            newHeaders.set('x-ratelimit-limit-tokens', String(rl.tokensLimit));
        }
        if (rl.tokensRemaining !== undefined) {
            newHeaders.set('x-ratelimit-remaining-tokens', String(rl.tokensRemaining));
        }
        if (rl.tokensReset) {
            newHeaders.set('x-ratelimit-reset-tokens', rl.tokensReset);
        }

        return new Response(response.body, {
            status: response.status,
            statusText: response.statusText,
            headers: newHeaders,
        });
    };
}

// ============================================================================
// Middleware Composition
// ============================================================================

/**
 * Composes multiple middleware into a single middleware.
 * Middleware are applied in order (first middleware wraps outermost).
 */
export function composeMiddleware(...middlewares: HttpMiddleware[]): HttpMiddleware {
    return (handler) => {
        return middlewares.reduceRight((h, mw) => mw(h), handler);
    };
}

/**
 * Creates a standard middleware stack.
 */
export function createStandardMiddleware(options: {
    logger: Logger;
    timeoutMs?: number;
}): HttpMiddleware {
    const middlewares: HttpMiddleware[] = [
        requestIdMiddleware(),
        loggingMiddleware({ logger: options.logger }),
        rateLimitMiddleware(),
    ];

    if (options.timeoutMs) {
        middlewares.push(timeoutMiddleware(options.timeoutMs));
    }

    return composeMiddleware(...middlewares);
}
